File: qt/gen_qactiongroup.go
package qt

import (
	"unsafe"
)

type QActionGroup__ExclusionPolicy int

const (
	QActionGroup__None              QActionGroup__ExclusionPolicy = 0
	QActionGroup__Exclusive         QActionGroup__ExclusionPolicy = 1
	QActionGroup__ExclusiveOptional QActionGroup__ExclusionPolicy = 2
)

type QActionGroup struct {
	h          uintptr
	isSubclass bool
}

// NewQActionGroup constructs a new QActionGroup object.
func NewQActionGroup(parent *QObject) *QActionGroup {
	ret := newQActionGroup(QActionGroup_new(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

func (this *QActionGroup) MetaObject() *QMetaObject {
	return newQMetaObject(QActionGroup_MetaObject(this.h))
}

func (this *QActionGroup) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QActionGroup_Metacast(this.h, param1_Cstring))
}

func QActionGroup_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QActionGroup_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QActionGroup) AddAction(a *QAction) *QAction {
	return newQAction(QActionGroup_AddAction(this.h, a.cPointer()))
}

func (this *QActionGroup) AddActionWithText(text string) *QAction {
	text_ms := struct_miqt_string{}
	text_ms.data = CString(text)
	text_ms.len = size_t(len(text))
	defer free(unsafe.Pointer(text_ms.data))
	return newQAction(QActionGroup_AddActionWithText(this.h, text_ms))
}

func (this *QActionGroup) AddAction2(icon *QIcon, text string) *QAction {
	text_ms := struct_miqt_string{}
	text_ms.data = CString(text)
	text_ms.len = size_t(len(text))
	defer free(unsafe.Pointer(text_ms.data))
	return newQAction(QActionGroup_AddAction2(this.h, icon.cPointer(), text_ms))
}

func (this *QActionGroup) RemoveAction(a *QAction) {
	QActionGroup_RemoveAction(this.h, a.cPointer())
}

func (this *QActionGroup) Actions() []*QAction {
	var _ma struct_miqt_array = QActionGroup_Actions(this.h)
	_ret := make([]*QAction, int(_ma.len))
	_outCast := (*[0xffff]*QAction)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		_ret[i] = newQAction(_outCast[i])
	}
	return _ret
}

func (this *QActionGroup) CheckedAction() *QAction {
	return newQAction(QActionGroup_CheckedAction(this.h))
}

func (this *QActionGroup) IsExclusive() bool {
	return (bool)(QActionGroup_IsExclusive(this.h))
}

func (this *QActionGroup) IsEnabled() bool {
	return (bool)(QActionGroup_IsEnabled(this.h))
}

func (this *QActionGroup) IsVisible() bool {
	return (bool)(QActionGroup_IsVisible(this.h))
}

func (this *QActionGroup) ExclusionPolicy() ExclusionPolicy {
	xxxxxxxxx
}

func (this *QActionGroup) SetEnabled(enabled bool) {
	QActionGroup_SetEnabled(this.h, (bool)(enabled))
}

func (this *QActionGroup) SetDisabled(b bool) {
	QActionGroup_SetDisabled(this.h, (bool)(b))
}

func (this *QActionGroup) SetVisible(visible bool) {
	QActionGroup_SetVisible(this.h, (bool)(visible))
}

func (this *QActionGroup) SetExclusive(exclusive bool) {
	QActionGroup_SetExclusive(this.h, (bool)(exclusive))
}

func (this *QActionGroup) SetExclusionPolicy(policy ExclusionPolicy) {
	QActionGroup_SetExclusionPolicy(this.h, policy)
}

func (this *QActionGroup) Triggered(param1 *QAction) {
	QActionGroup_Triggered(this.h, param1.cPointer())
}

func (this *QActionGroup) OnTriggered(slot func(param1 *QAction)) {
	QActionGroup_connect_Triggered(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QActionGroup_Triggered
func miqt_exec_callback_QActionGroup_Triggered(cb intptr_t, param1 *QAction) {
	gofunc, ok := cgo.Handle(cb).Value().(func(param1 *QAction))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQAction(param1)

	gofunc(slotval1)
}

func (this *QActionGroup) Hovered(param1 *QAction) {
	QActionGroup_Hovered(this.h, param1.cPointer())
}

func (this *QActionGroup) OnHovered(slot func(param1 *QAction)) {
	QActionGroup_connect_Hovered(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QActionGroup_Hovered
func miqt_exec_callback_QActionGroup_Hovered(cb intptr_t, param1 *QAction) {
	gofunc, ok := cgo.Handle(cb).Value().(func(param1 *QAction))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQAction(param1)

	gofunc(slotval1)
}

func QActionGroup_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QActionGroup_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QActionGroup_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QActionGroup_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QActionGroup) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QActionGroup_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QActionGroup) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QActionGroup_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QActionGroup_MetaObject
func miqt_exec_callback_QActionGroup_MetaObject(self QActionGroup, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QActionGroup{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QActionGroup) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QActionGroup_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QActionGroup) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QActionGroup_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QActionGroup_Metacast
func miqt_exec_callback_QActionGroup_Metacast(self QActionGroup, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QActionGroup{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/gen_qactiongroup.cpp
// +build ignore

#include <QAction>
#include <QActionGroup>
#include <QIcon>
#include <QList>
#include <QMetaObject>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qactiongroup.h>
#include "gen_qactiongroup.h"
class MiqtVirtualQActionGroup : public virtual QActionGroup {
public:
MiqtVirtualQActionGroup(QObject* parent): QActionGroup(parent) {};

virtual ~MiqtVirtualQActionGroup() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QActionGroup::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QActionGroup_MetaObject(const_cast<MiqtVirtualQActionGroup*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QActionGroup::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QActionGroup::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QActionGroup_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QActionGroup::qt_metacast(param1);

	}
};
QActionGroup* QActionGroup_new(QObject* parent) {
return new MiqtVirtualQActionGroup(parent);
}
void QActionGroup_virtbase(QActionGroup* src
, QObject** outptr_QObject
) {
*outptr_QObject = static_cast<QObject*>(src);
}
QMetaObject* QActionGroup_MetaObject(const QActionGroup* self) {
	return (QMetaObject*) self->metaObject();
}
void* QActionGroup_Metacast(QActionGroup* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QActionGroup_Tr(const char* s) {
	QString _ret = QActionGroup::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
QAction* QActionGroup_AddAction(QActionGroup* self, QAction* a) {
	return self->addAction(a);
}
QAction* QActionGroup_AddActionWithText(QActionGroup* self, struct miqt_string text) {
	QString text_QString = QString::fromUtf8(text.data, text.len);
	return self->addAction(text_QString);
}
QAction* QActionGroup_AddAction2(QActionGroup* self, QIcon* icon, struct miqt_string text) {
	QString text_QString = QString::fromUtf8(text.data, text.len);
	return self->addAction(*icon, text_QString);
}
void QActionGroup_RemoveAction(QActionGroup* self, QAction* a) {
	self->removeAction(a);
}
struct miqt_array /* of QAction* */  QActionGroup_Actions(const QActionGroup* self) {
	QList<QAction *> _ret = self->actions();
	// Convert QList<> from C++ memory to manually-managed C memory
	QAction** _arr = static_cast<QAction**>(malloc(sizeof(QAction*) * _ret.length()));
	for (size_t i = 0, e = _ret.length(); i < e; ++i) {
		_arr[i] = _ret[i];
	}
	struct miqt_array _out;
	_out.len = _ret.length();
	_out.data = static_cast<void*>(_arr);
	return _out;
}
QAction* QActionGroup_CheckedAction(const QActionGroup* self) {
	return self->checkedAction();
}
bool QActionGroup_IsExclusive(const QActionGroup* self) {
	return self->isExclusive();
}
bool QActionGroup_IsEnabled(const QActionGroup* self) {
	return self->isEnabled();
}
bool QActionGroup_IsVisible(const QActionGroup* self) {
	return self->isVisible();
}
ExclusionPolicy QActionGroup_ExclusionPolicy(const QActionGroup* self) {
	return self->exclusionPolicy();
}
void QActionGroup_SetEnabled(QActionGroup* self, bool enabled) {
	self->setEnabled(enabled);
}
void QActionGroup_SetDisabled(QActionGroup* self, bool b) {
	self->setDisabled(b);
}
void QActionGroup_SetVisible(QActionGroup* self, bool visible) {
	self->setVisible(visible);
}
void QActionGroup_SetExclusive(QActionGroup* self, bool exclusive) {
	self->setExclusive(exclusive);
}
void QActionGroup_SetExclusionPolicy(QActionGroup* self, ExclusionPolicy policy) {
	self->setExclusionPolicy(policy);
}
void QActionGroup_Triggered(QActionGroup* self, QAction* param1) {
	self->triggered(param1);
}
void QActionGroup_connect_Triggered(QActionGroup* self, intptr_t slot) {
	MiqtVirtualQActionGroup::connect(self, static_cast<void (QActionGroup::*)(QAction*)>(&QActionGroup::triggered), self, [=](QAction* param1) {
		QAction* sigval1 = param1;
		miqt_exec_callback_QActionGroup_Triggered(slot, sigval1);
	});
}
void QActionGroup_Hovered(QActionGroup* self, QAction* param1) {
	self->hovered(param1);
}
void QActionGroup_connect_Hovered(QActionGroup* self, intptr_t slot) {
	MiqtVirtualQActionGroup::connect(self, static_cast<void (QActionGroup::*)(QAction*)>(&QActionGroup::hovered), self, [=](QAction* param1) {
		QAction* sigval1 = param1;
		miqt_exec_callback_QActionGroup_Hovered(slot, sigval1);
	});
}
struct miqt_string QActionGroup_Tr2(const char* s, const char* c) {
	QString _ret = QActionGroup::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QActionGroup_Tr3(const char* s, const char* c, int n) {
	QString _ret = QActionGroup::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QActionGroup_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQActionGroup*>( (QActionGroup*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QActionGroup_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQActionGroup*)(self) )->virtualbase_MetaObject();
}
void QActionGroup_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQActionGroup*>( (QActionGroup*)(self) )->handle__Metacast = slot;
}
void* QActionGroup_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQActionGroup*)(self) )->virtualbase_Metacast(param1);
}
void QActionGroup_Delete(QActionGroup* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQActionGroup*>( self );
	} else {
		delete self;
	}
}
